linker/validate.go

```diff
@@ -359,3 +597,295 @@ func enumValCamelCase(name string) string {
 	}
 	return string(js)
 }
+
+func isBuiltinTypeName(typeName string) bool {
+	switch typeName {
+	case "int32", "int64", "uint32", "uint64", "sint32", "sint64",
+		"fixed32", "fixed64", "sfixed32", "sfixed64",
+		"bool", "double", "float", "string", "bytes":
+		return true
+	default:
+		return false
+	}
+}
+
+func getTypeName(fd protoreflect.FieldDescriptor) string {
+	switch fd.Kind() {
+	case protoreflect.MessageKind, protoreflect.GroupKind:
+		return "." + string(fd.Message().FullName())
+	case protoreflect.EnumKind:
+		return "." + string(fd.Enum().FullName())
+	default:
+		return fd.Kind().String()
+	}
+}
+
+func findExtensionRangeOptionSpan(
+	file protoreflect.FileDescriptor,
+	extended protoreflect.MessageDescriptor,
+	extRangeIndex int,
+	extRange *descriptorpb.DescriptorProto_ExtensionRange,
+	path ...int32,
+) (ast.SourceSpan, bool) {
+	// NB: Typically, we have an AST for a file and NOT source code info, because the
+	// compiler validates options before computing source code info. However, we might
+	// be validating an extension (whose source/AST we have), but whose extendee (and
+	// thus extension range options for declarations) could be in some other file, which
+	// could be provided to the compiler as an already-compiled descriptor. So this
+	// function can fallback to using source code info if an AST is not available.
+
+	if r, ok := file.(Result); ok && r.AST() != nil {
+		// Find the location using the AST, which will generally be higher fidelity
+		// than what we might find in a file descriptor's source code info.
+		exts := r.ExtensionsNode(extRange)
+		return findOptionSpan(r.FileNode(), exts, ast.AsNodeWithOptions(exts), extRange.Options.ProtoReflect().Descriptor(), path)
```

Comment:
It makes sense that normal and compact option statements are distinct in the AST, but it looks like it would be convenient to have an accessor for the contained option nodes that disambiguates. 

Wouldn't it be possible to have just `NodeWithOptions`, and let it range over compact option nodes for fields (etc.) and over normal option nodes for messages (etc.)?

Perhaps @emcfarlane can chime in with his perspective.

Reply:
I thought of another possible way to handle this, but it would be a backwards-incompatible change, so I'm still leaning towards what's here.

That idea is to change `GroupNode` so it _does not_ implement both `FieldDeclNode` and `MessageDeclNode`. Instead, it would _only_ implement `FieldDeclNode` and then have an `AsMessage() ast.MessageDeclNode` method. That method could return a new `*ast.GroupAsMessageNode` that would be defined like `type GroupAsMessageNode GroupNode`, and just have the method set for `MessageDeclNode`. In particular, its `RangeOptions` would iterate over the non-compact message options, where `GroupNode.RangeOptions` would range over the compact field options.

That idea would allow us to have a single `ast.NodeWithOptions` used for both compact and non-compact option declarations.

It's a breaking change to the `GroupNode` type, but it may be very unlikely to impact any existing code, so maybe it's worth the change to make this options stuff simpler? @timostamm, WDYT?

Reply:
(FWIW, this PR already does contain several other backwards-incompatible changes. But they are only incompatible for code that includes their own _implementations_ of some of the interfaces, and are perfectly compatible for code that merely _uses_ them. But maybe that's sufficient enough that it's worth making this other breaking change.)

Reply:
Okay, I pushed a commit that makes this change. Hopefully things look better now. both `*ast.GroupNode` and `*ast.MapFieldNode` now have an `AsMessage` accessor method to return a view of those nodes as a message. So, particularly for groups, the `GroupNode.RangeOptions` ranges through compact (field) options, and then `GroupNode.AsMessage().RangeOptions` ranges through non-compact (message) options.

Reply:
Oh, the synthetic nodes are an interesting solution. This clicks immediately for me, and I assume it will for others familiar with descriptors.